src/library_async.js

```diff
@@ -74,12 +74,17 @@ mergeInto(LibraryManager.library, {
               try {
                 return original.apply(null, arguments);
               } finally {
-                // Only functions in the list of known relevant imports are allowed to change the state.
-                // Note that invoke_* functions are allowed to change the state if we do not ignore
-                // indirect calls.
+                // Only functions in the list of known relevant imports are
+                // allowed to change the state.
                 if (Asyncify.state !== originalAsyncifyState &&
+                    // Changing the state from normal to disabled is allowed, as
+                    // that is what shutdown does.
+                    !(originalAsyncifyState === Asyncify.State.Normal &&
+                      Asyncify.state        === Asyncify.State.Disabled) &&
                     ASYNCIFY_IMPORTS.indexOf(x) < 0 &&
                     !x.startsWith('__asyncjs__') &&
+                    // invoke_* functions are allowed to change the state if we
+                    // do not ignore indirect calls.
                     !(x.startsWith('invoke_') && {{{ !ASYNCIFY_IGNORE_INDIRECT }}})) {
```

Comment:
This if condition is starting to get a little out of hand.  I won't block this change but if you can think of a way to refactor that would make this more readable I think that would be good.